executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -4,47 +4,119 @@ use std::{
-#[derive(Clone)]
-pub struct FinalizedRow<T> {
-    /// The values, indices according to the columns stored in [FinalizableData].
-    values: Vec<T>,
-    /// A bit vector indicating which cells are known. Values of unknown cells should be ignored.
+struct CompactData<T: FieldElement> {
+    /// The ID of the first column used in the table.
+    first_column_id: u64,
+    /// The length of a row in the table.
+    column_count: usize,
+    /// The cell values, stored in row-major order.
+    data: Vec<T>,
+    /// Bit vector of known cells, stored in row-major order.
     known_cells: BitVec,
 }
 
-impl<T> FinalizedRow<T> {
-    pub fn new(values: Vec<T>, known_cells: BitVec) -> Self {
+impl<T: FieldElement> CompactData<T> {
+    /// Creates a new empty compact data storage.
+    fn new(column_ids: &[PolyID]) -> Self {
+        let col_id_range = column_ids.iter().map(|id| id.id).minmax();
+        let (first_column_id, last_column_id) = col_id_range.into_option().unwrap();
         Self {
-            values,
-            known_cells,
+            first_column_id,
+            column_count: (last_column_id - first_column_id + 1) as usize,
+            data: Vec::new(),
+            known_cells: BitVec::new(),
+        }
+    }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    /// Returns the number of stored rows.
+    #[inline]
+    fn len(&self) -> usize {
+        self.data.len() / self.column_count
+    }
+
+    /// Truncates the data to `len` rows.
+    fn truncate(&mut self, len: usize) {
+        self.data.truncate(len * self.column_count);
+        self.known_cells.truncate(len * self.column_count);
+    }
+
+    fn clear(&mut self) {
+        self.data.clear();
+        self.known_cells.clear();
+    }
+
+    /// Appends a non-finalized row to the data, turning it into a finalized row.
+    #[inline]
+    fn push(&mut self, row: Row<T>) {
+        self.data.reserve(self.data.len() + self.column_count);
+        self.known_cells
+            .reserve(self.known_cells.len() + self.column_count);
+        for col_id in self.first_column_id..(self.first_column_id + self.column_count as u64) {
+            if let Some(v) = row.value(&PolyID {
+                id: col_id,
+                ptype: PolynomialType::Committed,
+            }) {
+                self.data.push(v);
+                self.known_cells.push(true);
+            } else {
+                self.data.push(T::zero());
+                self.known_cells.push(false);
+            }
         }
     }
+
+    #[inline]
+    fn try_remove_last_row(&mut self) -> bool {
+        if self.data.len() < self.column_count {
+            false
+        } else {
+            self.data.truncate(self.data.len() - self.column_count);
+            self.known_cells
+                .truncate(self.known_cells.len() - self.column_count);
+            true
+        }
+    }
+
+    #[inline]
+    fn get(&self, row: usize, col: u64) -> (T, bool) {
```

Comment:
I really would like to avoid any complication in the interface here. If we pass a PolID it suggests that we can also query for fixed columns which is just not true.

Reply:
Hmm, but now nothing prevents a caller from passing `some_fixed_poly_id.id`. If we received a `PolyID`, we could add an assertion that it is for a committed polynomial.